qiskit/utils/deprecation.py

```diff
@@ -89,27 +102,96 @@ def wrapper(*args, **kwargs):
-                    stacklevel=3,
-                )
-            else:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
-
-                kwargs[new_arg] = kwargs.pop(old_arg)
+        if old_arg not in kwargs:
+            continue
+        if new_arg in kwargs:
+            raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+        warnings.warn(old_kwarg_to_msg[old_arg], category=category, stacklevel=3)
+        if new_arg is not None:
+            kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _add_deprecation_to_docstring(
+    func: Callable, msg: str, *, since: Optional[str], pending: bool
+) -> None:
+    if "\n" in msg:
+        raise ValueError(
+            "Deprecation messages cannot contain new lines (`\\n`), but the deprecation for "
+            f'{func.__qualname__} had them. Usually this happens when using `"""` multiline '
+            f"strings; instead, use string concatenation.\n\n"
+            "This is a simplification to facilitate deprecation messages being added to our "
+            "documentation. If you have a compelling reason to need "
+            "new lines, feel free to improve this function or open a request at "
+            "https://github.com/Qiskit/qiskit-terra/issues."
+        )
+
+    if since is None:
+        version_str = "unknown"
+    else:
+        version_str = f"{since}_pending" if pending else since
+
+    # We insert deprecations in-between the description and the meta sections. The below is from
+    # https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections.
+    # We only include what is used by Qiskit projects for functions/methods to avoid wasting
+    # computation.
+    meta_lines = {
+        "Args:",
+        "Arguments:",
+        "Example:",
+        "Examples:",
+        "Note:",
+        "Notes:",
+        "Parameters:",
+        "Return:",
+        "Returns:",
+        "Raises:",
+        "References:",
+        "Yield:",
+        "Yields:",
+    }
+
+    indent = ""
+    meta_index = None
+    if func.__doc__:
+        original_lines = func.__doc__.splitlines()
+        content_encountered = False
+        for i, line in enumerate(original_lines):
+            stripped = line.strip()
+
+            # Determine the indent based on the first line with content. But, we don't consider the
+            # first line, which corresponds to the format """Docstring.""", as it does not properly
+            # capture the indentation of lines beneath it.
+            if not content_encountered and i != 0 and stripped:
+                num_spaces = len(line) - len(line.lstrip())
+                indent = " " * num_spaces
+                content_encountered = True
+
+            if stripped in meta_lines:
```

Comment:
We might want to normalise this to `if stripped.lower() in meta_lines` (and appropriate changes to `meta_lines`) to avoid an implicit fixing of the title-case/sentence-case discussion we had somewhere else, where napoleon doesn't make the same requirement.

Reply:
I instead added both the `See Also:` and `See also:` formats to the set. Only 4 of those necessary. That should be slightly less wasted computation than calling `.lower()` on every single line of the docstring.

Reply:
This is another case where, if we're intending this to be part of the public API for downstream projects, I'd think it's better to be more correct by default - having the section title capitalised is enforced by Terra's lint, but not necessarily anyone else's.  I wouldn't inherently worry about this timing - with all the other bits of the algorithm in place, calling `str.lower` is probably not going to have a meaningful impact.

As a brief illustration: by a very lazy `wc -l` on Terra, I found we have 223,314 lines in Python files within `qiskit` as of this PR.  If I generate that many 100-char lines (ignoring that only a tiny fraction of that are deprecated docstrings), and compare doing only the setup of this algorithm, not including the function wrapping, non-comprehension iteration, etc, the difference in runtime in adding a `lower` is about 14ms, and this is like the worst-case comparison:
```python
In [8]: import random
   ...: import string
   ...: docs = "\n".join("".join(random.choices(string.ascii_letters, k=100)) for _ in [None]*223_314)
   ...: test = frozenset(x.rstrip() for x in random.choices(docs.splitlines(), k=20))
   ...:
   ...: %timeit [s.strip() in test for s in docs.splitlines()]
   ...: %timeit [s.strip().lower() in test for s in docs.splitlines()]
50.4 ms ± 611 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
63.7 ms ± 529 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```